### ISSUE TITLE
Nginx memory leak (swaps server to death) when LDAP server is rebooted

### ISSUE THREAD

**Issue by mig5** [kvspb/nginx-auth-ldap#34]
Hi,

I originally discussed this bug in #33 , but it may be a separate issue.

I upgraded to Nginx 1.4.5 by re-compiling the source made available from http://dotdeb.org for Debian Squeeze and Wheezy, to include the latest version of your nginx-auth-ldap module.

I have been using nginx-auth-ldap module for a year or two, based on a checked out copy as at commit c2c3d5ea49191275da680b5e86108d6ad49b0aef, without issue

After upgrading to the latest version of your module, with 3 LDAP servers configured in our Nginx http { } section as follows:

```
ldap_server ldap1 {
  url ldaps://ldap1.example.com/dc=XXXXXXXX,dc=com?uid?sub?(objectClass=person);
  group_attribute memberUid; # default 'member'
  group_attribute_is_dn off;
  require valid_user;
        require group "cn=XXXXXXXX,ou=Groups,dc=XXXXXXXX,dc=com";
        satisfy any;
}

ldap_server ldap2 {
  url ldaps://ldap2.example.com/dc=XXXXXXXX,dc=com?uid?sub?(objectClass=person);
  group_attribute memberUid; # default 'member'
  group_attribute_is_dn off;
  require valid_user;
        require group "cn=XXXXXXXX,ou=Groups,dc=XXXXXXXX,dc=com";
        satisfy any;
}

ldap_server ldap3 {
  url ldaps://ldap3.example.com/dc=XXXXXXXX,dc=com?uid?sub?(objectClass=person);
  group_attribute memberUid; # default 'member'
  group_attribute_is_dn off;
  require valid_user;
        require group "cn=XXXXXXXX,ou=Groups,dc=XXXXXXXX,dc=com";
        satisfy any;
}

auth_ldap_cache_enabled on;
auth_ldap_cache_expiration_time 3600000;
auth_ldap_cache_size 1000;
```

... we found that if we shut down just one LDAP server, we got signal 9 errors in Nginx, and more brutally, a massive committed memory leak that would result in the servers swapping to death in a matter of minutes. (Graph below)

![screen shot 2014-02-21 at 3 32 16 pm](https://f.cloud.github.com/assets/99173/2242434/cd3b4b68-9d02-11e3-97b4-af122551ddf2.png)

Interestingly in further tests, it seems the memory leak starts to occur not _as_ an LDAP server goes offline, but as it comes back up, but is not yet ready to server connections or possibly if its firewall has not started.

Here is the debug of the log around that time:

2014/02/23 21:39:36 [info] 10292#0: http_auth_ldap: Could not connect
2014/02/23 21:39:36 [debug] 10288#0: event timer add: 0: 10000:1622182524
2014/02/23 21:39:36 [debug] 10290#0: worker cycle
2014/02/23 21:39:36 [debug] 10289#0: worker cycle
2014/02/23 21:39:36 [debug] 10288#0: http_auth_ldap: Connection scheduled for reconnection in 10000 ms
2014/02/23 21:39:36 [debug] 10290#0: accept mutex lock failed: 0
2014/02/23 21:39:36 [debug] 10292#0: http_auth_ldap: Closing connection (fd=43)
2014/02/23 21:39:36 [debug] 10289#0: accept mutex lock failed: 0
2014/02/23 21:39:36 [debug] 10290#0: epoll timer: 500
2014/02/23 21:39:36 [debug] 10288#0: timer delta: 4
2014/02/23 21:39:36 [debug] 10289#0: epoll timer: 500
2014/02/23 21:39:36 [debug] 10288#0: posted events 00000000
2014/02/23 21:39:36 [debug] 10292#0: event timer del: 43: 1622182521
2014/02/23 21:39:36 [debug] 10288#0: worker cycle
2014/02/23 21:39:36 [debug] 10292#0: reusable connection: 0
2014/02/23 21:39:36 [debug] 10288#0: accept mutex lock failed: 0
2014/02/23 21:39:36 [debug] 10292#0: event timer add: 0: 10000:1622182524
2014/02/23 21:39:36 [debug] 10288#0: epoll timer: 500
2014/02/23 21:39:36 [debug] 10292#0: http_auth_ldap: Connection scheduled for reconnection in 10000 ms
2014/02/23 21:39:36 [debug] 10292#0: timer delta: 3
2014/02/23 21:39:36 [debug] 10292#0: posted events 00000000
2014/02/23 21:39:36 [debug] 10292#0: worker cycle
2014/02/23 21:39:36 [debug] 10292#0: accept mutex lock failed: 0
2014/02/23 21:39:36 [debug] 10292#0: epoll timer: 500
2014/02/23 21:39:36 [debug] 10291#0: timer delta: 10010
2014/02/23 21:39:36 [debug] 10291#0: event timer del: 0: 1622172542
2014/02/23 21:39:36 [debug] 10291#0: http_auth_ldap: Connecting to LDAP server “XX.XX.XX.XX:636".
2014/02/23 21:39:36 [debug] 10291#0: socket 45
2014/02/23 21:39:36 [debug] 10291#0: epoll add connection: fd:45 ev:80000005
2014/02/23 21:39:36 [debug] 10291#0: connect to XX.XX.XX.XX:636, fd:45 #41
2014/02/23 21:39:36 [debug] 10291#0: http_auth_ldap: ngx_event_connect_peer() -> -2.
2014/02/23 21:39:36 [debug] 10291#0: event timer add: 45: 10000:1622182552
2014/02/23 21:39:36 [debug] 10291#0: posted events 00000000
2014/02/23 21:39:36 [debug] 10291#0: worker cycle
2014/02/23 21:39:36 [debug] 10291#0: accept mutex locked
2014/02/23 21:39:36 [debug] 10291#0: epoll timer: 10000
2014/02/23 21:39:36 [debug] 10291#0: epoll: fd:45 ev:001D d:095ECAE4
2014/02/23 21:39:36 [debug] 10291#0: epoll_wait() error on fd:45 ev:001D
2014/02/23 21:39:36 [debug] 10291#0: post event 09605A5C
2014/02/23 21:39:36 [debug] 10291#0: post event 09612A64
2014/02/23 21:39:36 [debug] 10291#0: timer delta: 1
2014/02/23 21:39:36 [debug] 10291#0: posted events 09612A64
2014/02/23 21:39:36 [debug] 10291#0: posted event 09612A64
2014/02/23 21:39:36 [debug] 10291#0: delete posted event 09612A64
2014/02/23 21:39:36 [debug] 10291#0: http_auth_ldap: Connect handler
2014/02/23 21:39:36 [debug] 10291#0: SSL_do_handshake: -1
2014/02/23 21:39:36 [debug] 10291#0: SSL_get_error: 5
2014/02/23 21:39:36 [info] 10291#0: peer closed connection in SSL handshake (111: Connection refused)
2014/02/23 21:39:36 [debug] 10291#0: SSL_do_handshake: -1
2014/02/23 21:39:36 [debug] 10291#0: SSL_get_error: 5
2014/02/23 21:39:36 [info] 10291#0: peer closed connection in SSL handshake (32: Broken pipe)
2014/02/23 21:39:36 [debug] 10291#0: SSL_do_handshake: -1
2014/02/23 21:39:36 [debug] 10291#0: SSL_get_error: 5
2014/02/23 21:39:36 [info] 10291#0: peer closed connection in SSL handshake (32: Broken pipe)
2014/02/23 21:39:36 [debug] 10291#0: SSL_do_handshake: -1
2014/02/23 21:39:36 [debug] 10291#0: SSL_get_error: 5
2014/02/23 21:39:36 [info] 10291#0: peer closed connection in SSL handshake (32: Broken pipe)
2014/02/23 21:39:36 [debug] 10291#0: SSL_do_handshake: -1
2014/02/23 21:39:36 [debug] 10291#0: SSL_get_error: 5
2014/02/23 21:39:36 [info] 10291#0: peer closed connection in SSL handshake (32: Broken pipe)
2014/02/23 21:39:36 [debug] 10291#0: SSL_do_handshake: -1
2014/02/23 21:39:36 [debug] 10291#0: SSL_get_error: 5

It is these SSL_do_handshake/SSL_get_error / Broken pipe messages that suddenly start appearing at hundreds per second in the log, coinciding with a massive spike in memory. While the LDAP server is still rebooting, we see 'couldn't connect to LDAP server' or similar messages, in a 10000ms loop, but they don't crash or cause memory leaks, only once the other SSL handshake messages appear.

This occurs even if there are no virtualhosts using the LDAP config at all in their server { } block, just the main config in http { }.

Per the other ticket #33, where he links to a specific commit he is using that relates to event handlers and SSL. I tried reverting that commit and recompiling, but I still reproduced.

I wonder if it's related to the SSL work done in commit f4d1da9bb71e314fb116229ec2a0617e175349a5, as I see changes there related to SSL handshake etc, but I didn't try rolling back as it was more critical for me to revert back to a working state (this bug crashed 50 servers when we rebooted our ldap3 server :( ) I'm also not familiar enough with C to identify any memory leak issue there.

I ultimately reverted back to the commit c2c3d5ea49191275da680b5e86108d6ad49b0aef, and I don't reproduce the issue when rebooting the (single) LDAP server it uses.

Our LDAP servers are also Debian Wheezy, OpenLDAP version 2.4.31-1+nmu2

Let me know if I can provide any other info that might be helpful in debugging
